From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 5 Apr 2025 13:37:48 +0200
Subject: [PATCH 018/225] iommu/sprd: add support for reserved regions
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add support for initial mappings defined in the device tree. Since these
are set up before the device is attached, move the page table allocation
to a separate "finalize" function that is invoked as soon as a device is
available.

Initial mappings (e.g. for the boot framebuffer) often require the
virtual address to match the physical address. To allow configuring the
IOMMU's virtual address range, read the new "sprd,iova-range" property
from the device tree.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/iommu/sprd-iommu.c | 84 +++++++++++++++++++++++++++++---------
 1 file changed, 65 insertions(+), 19 deletions(-)

diff --git a/drivers/iommu/sprd-iommu.c b/drivers/iommu/sprd-iommu.c
index c1a34445d..4cd7bc2e6 100644
--- a/drivers/iommu/sprd-iommu.c
+++ b/drivers/iommu/sprd-iommu.c
@@ -13,6 +13,7 @@
 #include <linux/iommu.h>
 #include <linux/mfd/syscon.h>
 #include <linux/module.h>
+#include <linux/of_iommu.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
 #include <linux/regmap.h>
@@ -67,6 +68,8 @@ struct sprd_iommu_device {
 	enum sprd_iommu_version	ver;
 	u32			*prot_page_va;
 	dma_addr_t		prot_page_pa;
+	u32			iova_start;
+	u32			iova_end;
 	void __iomem		*base;
 	struct device		*dev;
 	struct iommu_device	iommu;
@@ -79,6 +82,7 @@ struct sprd_iommu_domain {
 	u32			*pgt_va; /* page table virtual address base */
 	dma_addr_t		pgt_pa; /* page table physical address base */
 	struct sprd_iommu_device	*sdev;
+	bool			attached;
 };
 
 static const struct iommu_ops sprd_iommu_ops;
@@ -133,9 +137,32 @@ sprd_iommu_pgt_size(struct iommu_domain *domain)
 		SPRD_IOMMU_PAGE_SHIFT) * sizeof(u32);
 }
 
+static int sprd_iommu_finalize_domain(struct sprd_iommu_domain *dom,
+				      struct sprd_iommu_device *sdev)
+{
+	size_t pgt_size;
+
+	dom->sdev = sdev;
+
+	dom->domain.pgsize_bitmap = SPRD_IOMMU_PAGE_SIZE;
+
+	dom->domain.geometry.aperture_start = sdev->iova_start;
+	dom->domain.geometry.aperture_end = sdev->iova_end;
+	dom->domain.geometry.force_aperture = true;
+
+	pgt_size = sprd_iommu_pgt_size(&dom->domain);
+	dom->pgt_va = dma_alloc_coherent(sdev->dev, pgt_size, &dom->pgt_pa,
+					 GFP_KERNEL);
+	if (!dom->pgt_va)
+		return -ENOMEM;
+
+	return 0;
+}
+
 static struct iommu_domain *sprd_iommu_domain_alloc_paging(struct device *dev)
 {
 	struct sprd_iommu_domain *dom;
+	int ret;
 
 	dom = kzalloc_obj(*dom);
 	if (!dom)
@@ -143,11 +170,15 @@ static struct iommu_domain *sprd_iommu_domain_alloc_paging(struct device *dev)
 
 	spin_lock_init(&dom->pgtlock);
 
-	dom->domain.pgsize_bitmap = SPRD_IOMMU_PAGE_SIZE;
+	if (dev) {
+		struct sprd_iommu_device *sdev = dev_iommu_priv_get(dev);
 
-	dom->domain.geometry.aperture_start = 0;
-	dom->domain.geometry.aperture_end = SZ_256M - 1;
-	dom->domain.geometry.force_aperture = true;
+		ret = sprd_iommu_finalize_domain(dom, sdev);
+		if (ret) {
+			kfree(dom);
+			return NULL;
+		}
+	}
 
 	return &dom->domain;
 }
@@ -226,23 +257,21 @@ static void sprd_iommu_hw_en(struct sprd_iommu_device *sdev, bool en)
 
 static void sprd_iommu_cleanup(struct sprd_iommu_domain *dom)
 {
-	size_t pgt_size;
-
 	/* Nothing need to do if the domain hasn't been attached */
-	if (!dom->sdev)
+	if (!dom->sdev || dom->sdev->dom != dom)
 		return;
 
-	pgt_size = sprd_iommu_pgt_size(&dom->domain);
-	dma_free_coherent(dom->sdev->dev, pgt_size, dom->pgt_va, dom->pgt_pa);
 	sprd_iommu_hw_en(dom->sdev, false);
-	dom->sdev = NULL;
+	dom->sdev->dom = NULL;
 }
 
 static void sprd_iommu_domain_free(struct iommu_domain *domain)
 {
 	struct sprd_iommu_domain *dom = to_sprd_domain(domain);
+	size_t pgt_size = sprd_iommu_pgt_size(domain);
 
 	sprd_iommu_cleanup(dom);
+	dma_free_coherent(dom->sdev->dev, pgt_size, dom->pgt_va, dom->pgt_pa);
 	kfree(dom);
 }
 
@@ -252,19 +281,16 @@ static int sprd_iommu_attach_device(struct iommu_domain *domain,
 {
 	struct sprd_iommu_device *sdev = dev_iommu_priv_get(dev);
 	struct sprd_iommu_domain *dom = to_sprd_domain(domain);
-	size_t pgt_size = sprd_iommu_pgt_size(domain);
+	int ret;
 
 	/* The device is attached to this domain */
 	if (sdev->dom == dom)
 		return 0;
 
-	/* The first time that domain is attaching to a device */
-	if (!dom->pgt_va) {
-		dom->pgt_va = dma_alloc_coherent(sdev->dev, pgt_size, &dom->pgt_pa, GFP_KERNEL);
-		if (!dom->pgt_va)
-			return -ENOMEM;
-
-		dom->sdev = sdev;
+	if (!dom->sdev) {
+		ret = sprd_iommu_finalize_domain(dom, sdev);
+		if (ret)
+			return ret;
 	}
 
 	sdev->dom = dom;
@@ -298,7 +324,7 @@ static int sprd_iommu_map(struct iommu_domain *domain, unsigned long iova,
 	unsigned long end = domain->geometry.aperture_end;
 
 	if (!dom->sdev) {
-		pr_err("No sprd_iommu_device attached to the domain\n");
+		pr_err("sprd_iommu_domain is not finalized!\n");
 		return -EINVAL;
 	}
 
@@ -413,6 +439,7 @@ static const struct iommu_ops sprd_iommu_ops = {
 	.probe_device	= sprd_iommu_probe_device,
 	.device_group	= generic_single_device_group,
 	.of_xlate	= sprd_iommu_of_xlate,
+	.get_resv_regions = of_iommu_get_resv_regions,
 	.owner		= THIS_MODULE,
 	.default_domain_ops = &(const struct iommu_domain_ops) {
 		.attach_dev	= sprd_iommu_attach_device,
@@ -462,6 +489,7 @@ static int sprd_iommu_probe(struct platform_device *pdev)
 	struct sprd_iommu_device *sdev;
 	struct device *dev = &pdev->dev;
 	void __iomem *base;
+	u32 iova_range[2];
 	int ret;
 
 	sdev = devm_kzalloc(dev, sizeof(*sdev), GFP_KERNEL);
@@ -475,6 +503,24 @@ static int sprd_iommu_probe(struct platform_device *pdev)
 	}
 	sdev->base = base;
 
+	ret = of_property_read_u32_array(dev->of_node, "sprd,iova-range",
+					 iova_range, 2);
+	if (ret == -ENOENT) {
+		iova_range[0] = 0;
+		iova_range[1] = SZ_256M;
+	} else if (ret < 0) {
+		dev_err(dev, "Invalid sprd,iova-range property\n");
+		return PTR_ERR(base);
+	}
+
+	if (!IS_ALIGNED(iova_range[0], roundup_pow_of_two(iova_range[1]))) {
+		dev_err(dev, "Misaligned sprd,iova-range\n");
+		return PTR_ERR(base);
+	}
+
+	sdev->iova_start = iova_range[0];
+	sdev->iova_end = iova_range[0] + iova_range[1] - 1;
+
 	sdev->prot_page_va = dma_alloc_coherent(dev, SPRD_IOMMU_PAGE_SIZE,
 						&sdev->prot_page_pa, GFP_KERNEL);
 	if (!sdev->prot_page_va)
